tls/s2n_tls13_handshake.c

```diff
@@ -162,91 +162,177 @@ int s2n_tls13_compute_shared_secret(struct s2n_connection *conn, struct s2n_blob
 
     POSIX_GUARD_RESULT(s2n_connection_wipe_all_keyshares(conn));
 
+    /* It would make more sense to wipe the PSK secrets in s2n_tls13_handle_early_secret,
+     * but at that point we don't know whether or not the server will request a HRR request
+     * and we'll have to use the secrets again.
+     *
+     * Instead, wipe them here when we wipe all the other connection secrets. */
+    POSIX_GUARD_RESULT(s2n_psk_parameters_wipe_secrets(&conn->psk_params));
+
     return S2N_SUCCESS;
 }
 
-/*
- * This function executes after Server Hello is processed
- * and handshake hashes are computed. It produces and configure
- * the shared secret, handshake secrets, handshake traffic keys,
- * and finished keys.
- */
-int s2n_tls13_handle_handshake_secrets(struct s2n_connection *conn)
+int s2n_tls13_handle_early_secret(struct s2n_connection *conn)
 {
     POSIX_ENSURE_REF(conn);
+
+    struct s2n_psk *psk = conn->psk_params.chosen_psk;
+
+    /* If the client is sending early data, then the client will need to calculate
+     * the early secret before the server chooses a PSK. */
+    if (conn->mode == S2N_CLIENT && conn->early_data_state == S2N_EARLY_DATA_REQUESTED) {
+        POSIX_GUARD_RESULT(s2n_array_get(&conn->psk_params.psk_list, 0, (void**) &psk));
+        POSIX_ENSURE_REF(psk);
+
+        /* We need to set the early data cipher suite so that the secret derivation
+         * logic uses the right hmac algorithm.
+         */
+        POSIX_ENSURE_REF(psk->early_data_config.cipher_suite);
+        conn->secure.cipher_suite = psk->early_data_config.cipher_suite;
+    }
+
+    /* get tls13 key context */
+    s2n_tls13_connection_keys(secrets, conn);
+
+    /* derive early secrets */
+    POSIX_GUARD(s2n_tls13_derive_early_secret(&secrets, psk));
+
+    return S2N_SUCCESS;
+}
+
+
+int s2n_tls13_handle_early_traffic_secret(struct s2n_connection *conn)
+{
+    POSIX_ENSURE_REF(conn);
+
+    /* get tls13 key context */
+    s2n_tls13_connection_keys(secrets, conn);
+
+    struct s2n_hash_state hash_state = {0};
+    POSIX_GUARD(s2n_handshake_get_hash_state(conn, secrets.hash_algorithm, &hash_state));
+
+    s2n_stack_blob(early_traffic_secret, secrets.size, S2N_TLS13_SECRET_MAX_LEN);
+    POSIX_GUARD(s2n_tls13_derive_early_traffic_secret(&secrets, &hash_state, &early_traffic_secret));
+
+    /* trigger callbacks */
+    if (conn->secret_cb && conn->config->quic_enabled) {
+        POSIX_GUARD(conn->secret_cb(conn->secret_cb_context, conn, S2N_CLIENT_EARLY_TRAFFIC_SECRET,
+                early_traffic_secret.data, early_traffic_secret.size));
+    }
+    s2n_result_ignore(s2n_key_log_tls13_secret(conn, &early_traffic_secret, S2N_CLIENT_EARLY_TRAFFIC_SECRET));
+
+    /* produce traffic key */
+    s2n_tls13_key_blob(early_traffic_key, conn->secure.cipher_suite->record_alg->cipher->key_material_size);
+    struct s2n_blob app_iv = { .data = conn->secure.client_implicit_iv, .size = S2N_TLS13_FIXED_IV_LEN };
```

Comment:
Is it `S2N_TLS_MAX_IV_LEN == S2N_TLS13_FIXED_IV_LEN` true? Why not the following?
```c
struct s2n_blob app_iv = {0};
POSIX_GUARD(s2n_blob_init(&app_iv, conn->secure.client_implicit_iv, S2N_TLS13_FIXED_IV_LEN));
```

Reply:
S2N_TLS13_FIXED_IV_LEN < S2N_TLS_MAX_IV_LEN. S2N_TLS_MAX_IV_LEN is the size of the iv byte array stored on the connection, but TLS1.3 never uses the whole array.